recipe_recommender.mdx

```diff
@@ -0,0 +1,395 @@
+recipe that we will use throughout the project.
+
+Next, we need to set up the API keys for the services we will be using.
+
+### OpenAI setup
+1. Create an API key at https://platform.openai.com/account/api-keys
+2. Save it as environment variable under `OPENAI_API_KEY`
+
+### Pinecone setup
+1. Create an API key at https://app.pinecone.io/
+2. Save it as environment variable under `PINECONE_API_KEY`
+
+### PromptLayer setup
+1. Create an API key at https://promptlayer.com/home
+2. Save it as environment variable under `PROMPTLAYER_API_KEY`
+
+# Diving into the Code
+
+Start by creating a `settings.py` file for our application level constants.
+<CodeGroup>
+```python settings.py
+from pathlib import Path
+
+BASE_DIR = Path(__file__).resolve().parent
+DATA_DIR = BASE_DIR / 'data'  # directory to store our recipe text files
+EMBEDDING_MODEL = 'text-similarity-ada-001'  # OpenAI embedding model
+PINECONE_INDEX_NAME = 'recipes'  # name of our Pinecone index
+```
+</CodeGroup>
+
+Next, create a `openai_utils.py` file. This includes a `get_vector_for_text` function that allows us to pass in text
+and responds with a vector. We will later store these vectors in Pinecone.
+
+<CodeGroup>
+```python openai_utils.py
+import openai
+
+from settings import EMBEDDING_MODEL
+
+def get_vector_for_text(text):
+    response = openai.Embedding.create(
+        input=text,
+        engine=EMBEDDING_MODEL,
+    )
+    return response['data'][0]['embedding']
+```
+</CodeGroup>
+
+Now create a new file called `io_utils.py`. This file includes a helper function called `get_recipe_filepaths` to
+retrieve a list of recipe file paths.
+
+<CodeGroup>
+```python io_utils.py
+import os
+
+from settings import DATA_DIR
+
+def get_recipe_filepaths():
+    results = []
+
+    for file in os.listdir(DATA_DIR):
+        if file.endswith('.txt'):
+            results.append(DATA_DIR / file)
+
+    return results
+```
+</CodeGroup>
+
+Next, create a `files.py` file that contains the functionality used to parse file names and read text files. The
+file name (without the extension) acts as the unique ID for each recipe vector stored in Pinecone.
+
+<CodeGroup>
+```python files.py
+import os
+
+def extract_filename_without_extension(path):
+    """Extracts the filename without its extension from a given path"""
+    base_name = os.path.basename(path)
+    file_name_without_extension, _ = os.path.splitext(base_name)
+    return file_name_without_extension
+
+def read_txt_file(filepath):
+    """Reads the content of a text file and returns it as a string"""
+    with open(filepath, 'r') as file:
+        content = file.read()
+
+    return content
+```
+</CodeGroup>
+
+Create a `pinecone_populate.py` file next. Once created, run the script. During execution, the script first creates a
+Pinecone index. It then reads each of our recipes, generates a vector from it, and then inserts the vector as a new
+record in our Pinecone index.
+
+<CodeGroup>
```

Comment:
Might be clearer instead of having each CodeGroup be one whole file, to split up the CodeGroup and explain each section